File: src/main/java/victor/training/cleancode/GuardClauses.java
package victor.training.cleancode;

import lombok.Data;
import lombok.Value;

import java.util.ArrayList;
import java.util.List;

public class GuardClauses {

  public static final int DEAD_PAY_AMOUNT = 1;

  public int getPayAmount(Marine marine, BonusPackage bonusPackage) {
    int result;
    if (marine != null && (bonusPackage.getValue() > 100 || bonusPackage.getValue() < 10)) {
      if (!isDead(marine)) {
        if (!marine.isRetired()) {
          if (marine.getYearsService() != null) {
            result = marine.getYearsService() * 100 + bonusPackage.getValue();
            if (!marine.getAwards().isEmpty()) {
              result += 1000;
            }
            if (marine.getAwards().size() >= 3) {
              result += 2000;
            }
            // HEAVY core logic here, business-rules ...
          } else {
            throw new IllegalArgumentException("Any marine should have the years of service set");
          }
        } else result = retiredAmount();
      } else {
        result = DEAD_PAY_AMOUNT;
      }
    } else{
      throw new IllegalArgumentException("Not applicable!");
    }
    return result; // TODO ALT-ENTER move return closer
  }

  private boolean isDead(Marine marine) {
    return false;
  }

  private int retiredAmount() {
    return 2;
  }

}

@Value
class Marine {
  boolean dead;
  boolean retired;
  Integer yearsService;
  List<Award> awards = new ArrayList<>();

}

@Data
class BonusPackage {
  int value;
}

class Award {

}

File: src/main/java/victor/training/cleancode/GuardClauses.kt
package victor.training.cleancode

class GuardClauses {
    fun getPayAmount(marine: Marine): Int {
      /*  if (marine == null) {
            throw RuntimeException("Marine is null")
        }
        else*/ if (isDead(marine)) {
            return DEAD_PAY_AMOUNT
        }
        else if (marine.isRetired) return retiredAmount()
         // Principle of single return function
//        else if (marine.getYearsService() == null) {
//            throw IllegalArgumentException("Any marine should have the years of service set")
        else {
            var result: Int = marine.getYearsService()!! * 100
            if (marine.awards.isNotEmpty()) {
                result += 1000
            }
            if (marine.awards.size >= 3) {
                result += 2000
            }
            // HEAVY logic here...
            return result // TODO ALT-ENTER move return closer
         }
    }

    private fun isDead(marine: Marine): Boolean {
        return false
    }

    private fun retiredAmount(): Int {
        return 2
    }

    companion object {
        const val DEAD_PAY_AMOUNT = 1
    }
}

class Marine(val isDead: Boolean, val isRetired: Boolean, private val yearsService: Int) {
    val awards: List<Award> = ArrayList()

    fun getYearsService(): Int? {
        return yearsService
    }
}

 class Award